### PR TITLE
Implementation of Fortran create and free decomp

### PR DESCRIPTION
**Pull request by MiCurry**
These commits implements the Fortran side of the create and free decomp routines and defines  `SMIOLf_decomp` type which is interoperable with the C Struct SMIOL_decomp type. It also implements unit tests in smiol_runner.F90.

### CODE REVIEW

**Comment by MiCurry** [comment]
At the current moment, these commits fail for any arrays of size greater then 1.

**Comment by MiCurry** [comment]
@mgduda, I believe this PR is all good to go, but please let me know if you see or have any problems with it!

**Comment by mgduda** [comment]
Let's correct this new compiler warning:
```
smiolf.F90:531:59:

  531 |             function SMIOL_free_decomp(decomp) result(ierr) bind(C, name='SMIOL_free_decomp')
      |                                                           1
Warning: Variable ‘ierr’ at (1) may not be a C interoperable kind but it is BIND(C) [-Wc-binding-type]
```

**Comment by mgduda** [comment]
I think the type of the `compute_elements` and `io_elements` arrays in C code should be `int64_t` rather than `uint64_t`.

**Comment by mgduda** [comment]
For the unit tests, would it be worth testing that we get correct behavior when the number of elements is zero?

**Comment by mgduda** [comment]
In `smiol_runner.F90`, "assocaited" should be "associated".

**Comment by MiCurry** [comment]
Okay, those fixes have just been pushed up. I added a new commit to update the C code to use `int64_t` instead of `uint64_t`.

File: smiol_runner.F90 [hunk 1]
```diff
@@ -212,4 +243,138 @@ function test_init_finalize() result(ierrcount)
...
+        type (SMIOLf_decomp), pointer :: decomp => null()
+
+        write(0,'(a)') '********************************************************************************'
+        write(0,'(a)') '************ SMIOLf_create_decomp / SMIOLf_free_decomp tests *******************'
+        write(0,'(a)') ''
+
+        ierrcount = 0
+
+        ! Test with 0 elements
+        write(0,'(a)',advance='no') 'Everything OK for SMIOLf_create_decomp with 0 elements: '
```
Comment (mgduda) [inline smiol_runner.F90]: Should this say "with 1 element"?

Comment (mgduda) [reply smiol_runner.F90]: Alternatively, should `n_compute_elements` and `n_io_elements` be zero in the two lines below?

File: smiol_runner.F90 [hunk 2]
```diff
@@ -212,4 +243,138 @@ function test_init_finalize() result(ierrcount)
...
+        deallocate(io_elements)
+
+        ! Free Decomp
+        write(0,'(a)',advance='no') 'Everything OK for SMIOLf_free_decomp with 0 elements: '
+        ierr = SMIOLf_free_decomp(decomp)
+        if (ierr == SMIOL_SUCCESS .and. .not. associated(decomp)) then
+            write(0,'(a)') "PASS"
+        else if (ierr /= SMIOL_SUCCESS .and. associated(decomp)) then
+            write(0,'(a)') "FAIL - ierr did not return SMIOL_SUCCESS, and decomp was still associated"
+            ierr = ierrcount + 1
```
Comment (mgduda) [inline smiol_runner.F90]: I think this should be `ierrcount = ierrcount + 1` here and a few other places.

File: smiol_runner.F90 [hunk 3]
```diff
@@ -212,4 +243,138 @@ function test_init_finalize() result(ierrcount)
...
+        else if (ierr == SMIOL_SUCCESS .and. associated(decomp)) then
+            write(0,'(a)') "FAIL - ierr returned SMIOL_SUCCESS but the decomp was associated"
+            ierr = ierrcount + 1
+        endif
+
+        ! Small number of Compute and IO Elements
+        write(0,'(a)',advance='no') 'Everything OK for SMIOLf_create_decomp 1 element: '
+        n_compute_elements = 0
+        n_io_elements = 0
```
Comment (mgduda) [inline smiol_runner.F90]: Should `n_compute_elements` and `n_io_elements` be 1 here?

File: smiol_runner.F90 (hunk 2 above)
Comment (MiCurry) [reply smiol_runner.F90]: Yes, and some ierrcounts completely missing down below.

**Comment by MiCurry** [comment]
@mgduda thanks for seeing those! Those have been addressed in the latest update.

**Comment by mgduda** [comment]
@MiCurry I think we're getting close. Here are a few other points we might like to correct before merging this PR:
1. In `SMIOL_create_decomp` and `SMIOL_free_decomp`, the return value is set to 0 if there are no errors, and I think we should use `SMIOL_SUCCESS` instead
2. Commit c76fb808 doesn't compile, giving the following error:
```
smiol_runner.F90:46:8:

   46 |     if (SMIOLf_create_decomp(n_compute_elements, n_io_elements, compute_elements, io_elements, decomp) /= SMIOL_SUCCESS) then
      |        1
Error: Type mismatch in argument ‘n_compute_elements’ at (1); passed INTEGER(4) to INTEGER(8)
```
3. "succesfully" should be "successfully"

**Comment by MiCurry** [comment]
Sounds good, I'll fix those up and I'll let you know when things are updated.

@mgduda should I also update how I have declared Fortran interoperable C integers? In these commits they are: `integer(c_size_t)`, but should I update them to be `integer(kind=c_size_t)`?

**Comment by mgduda** [comment]
@MiCurry The Fortran 2003 standard writes interoperable integers as, e.g.,
```fortran
integer(c_int) :: i
```
So, I don't think we need to change `integer(c_size_t)` to `integer(kind=c_size_t)`. At some point in the near future, we can make a PR that makes the Fortran coding style consistent across all routines.

**Comment by MiCurry** [comment]
@mgduda sounds good. I've let the integer deceleration as they are, but I have updated the above changes you requested.

**Pull request merged**